### ISSUE TITLE
`let _ : <type> = ...` in Toplevel

### ISSUE THREAD

**Issue by NathanReb** [NathanReb/ocaml-native-toplevel-experience#2]
The native toplevel does not print evaluations for `let _ : <type> = ...` but the bytecode toplevel does. We'd like the former to align with the latter.

Here's an example from the [ocaml-jit's test suite](https://github.com/NathanReb/ocaml-jit/blob/main/tests/let_underscore.ml):
```ocaml
# let _ : int = 0;;
- : int = 0
```
Is what you would get in a bytecode toplevel. In a native toplevel, it would simply print nothing.

The bytecode toplevel is also partially affected by this bug, in some less trivial cases.

**Comment by NathanReb** [comment]
Next steps on this are:
1. [ ] investigate when this triggers in the bytecode toplevel
2. [ ] fix the native toplevel in [ocaml-flambda/flambda-backend](https://github.com/ocaml-flambda/flambda-backend)
3. [ ] report and fix it in upstream OCaml for both toplevels

**Comment by dra27** [comment]
The difference here is that the native toplevel is not re-writing the expression when it needs to (there's nothing triggering, per se, in the bytecode toplevel - it doesn't have to do this dance). This diff fixes trunk (prior to #1 being fixed):

```diff
diff --git a/toplevel/native/topeval.ml b/toplevel/native/topeval.ml
index 0d4cc77082..4a53031709 100644
--- a/toplevel/native/topeval.ml
+++ b/toplevel/native/topeval.ml
@@ -136,6 +136,16 @@ let execute_phrase print_outcome ppf phr =
             let pat = Ast_helper.Pat.var (Location.mknoloc "_$") in
             let vb = Ast_helper.Vb.mk ~loc ~attrs pat e in
             [ Ast_helper.Str.value ~loc Asttypes.Nonrecursive [vb] ], true
+        | [ { pstr_desc = Pstr_value (Asttypes.Nonrecursive,
+                                      [{ pvb_expr = e
+                                       ; pvb_pat = { ppat_desc = Ppat_constraint({ppat_desc = Ppat_any; _}, typ) ; _ }
+                                       ; pvb_attributes = attrs
+                                       ; _ }])
+            ; pstr_loc = loc }
+          ] ->
+            let pat = Ast_helper.Pat.(constraint_ (var (Location.mknoloc "_$")) typ) in
+            let vb = Ast_helper.Vb.mk ~loc ~attrs pat e in
+            [ Ast_helper.Str.value ~loc Asttypes.Nonrecursive [vb] ], true
         | _ -> sstr, false
       in
       let (str, sg, names, shape, newenv) =

```
I _think_ the fix for after #1 is fixed is slightly simpler:
```diff
diff --git a/toplevel/native/topeval.ml b/toplevel/native/topeval.ml
index b1f2bd9439..3fa0e31a21 100644
--- a/toplevel/native/topeval.ml
+++ b/toplevel/native/topeval.ml
@@ -188,8 +188,10 @@ let execute_phrase print_outcome ppf phr =
          | [ { str_desc = Tstr_value (Asttypes.Nonrecursive,
                                        [{ vb_expr = e
                                         ; vb_pat =
-                                            { pat_desc = Tpat_any;
-                                              pat_extra = []; _ }
+                                            { pat_desc = Tpat_any
+                                            ; pat_extra =
+                                                ([] | [Tpat_constraint _, _, _])
+                                            ; _ }
                                         ; vb_attributes = attrs }])
              ; str_loc = loc }
            ] ->

```
but it needs a type-system expert to confirm that that pattern match on `pat_extra` is correct.

**Comment by dra27** [comment]
This is fixed in 4.14!

**Comment by NathanReb** [comment]
Awesome!
